### PR TITLE
feat(docs): Typescript vuex smart module usage

### PR DESCRIPTION
**Pull request by Evertvdw**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [ ] Feature
- [x] Documentation
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch (or `v[X]` branch)
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] It's been tested on a Cordova (iOS, Android) app
- [ ] It's been tested on a Electron app
- [x] Any necessary documentation has been added or updated [in the docs](https://github.com/quasarframework/quasar/tree/dev/docs) (for faster update click on "Suggest an edit on GitHub" at bottom of page) or explained in the PR's description.

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

### CODE REVIEW

File: docs/src/pages/quasar-cli/vuex-store.md [hunk 1]
```diff
@@ -146,5 +146,304 @@ export default defineComponent({
 </script>
 ```
 
+::: warning
+With Vuex currently only the state is strongly typed. If you want to use typed getters/mutations/actions you will need to use either an extra package on top of Vuex or a replacement of Vuex.
```
Comment (smolinari) [inline docs/src/pages/quasar-cli/vuex-store.md]: Comma corrections

> With Vuex, currently, only 

> If you want to use type getters/mutations/actions,

File: docs/src/pages/quasar-cli/vuex-store.md [hunk 2]
```diff
@@ -146,5 +146,304 @@ export default defineComponent({
...
+:::
+
+### Using Vuex Smart Module
+One of the options for a fully typed store is a package called `vuex-smart-module`. You can add this package by running the following command:
+```bash
+yarn add vuex-smart-module
+```
+
+Once installed you need to edit your `src/store/index.ts` file to use this package to create the store. Edit your store index file to resemble the following:
```
Comment (smolinari) [inline docs/src/pages/quasar-cli/vuex-store.md]: Comma correction

> Once installed,

File: docs/src/pages/quasar-cli/vuex-store.md [hunk 3]
```diff
@@ -146,5 +146,304 @@ export default defineComponent({
...
+    // and other options, normally passed to Vuex `createStore`
+  });
+
+  return rootStore;
+});
+
+export const useStore = createComposable(root);
+```
+
+You can use modules just as with normal Vuex, and in that module you can choose to put everything in one file or use separate files for state, getters, mutations and actions. Or of course a combination of those two.
```
Comment (smolinari) [inline docs/src/pages/quasar-cli/vuex-store.md]: Sentence completion

> Or, of course,  you can use a combination of those two.

File: docs/src/pages/quasar-cli/vuex-store.md [hunk 4]
```diff
@@ -146,5 +146,304 @@ export default defineComponent({
...
+
+Just import the module in `src/store/index.ts` and add it to your `rootConfig`. For an example look [here](https://github.com/ktsn/vuex-smart-module#usage)
```
Comment (smolinari) [inline docs/src/pages/quasar-cli/vuex-store.md]: Comma correction

> For an example,

File: docs/src/pages/quasar-cli/vuex-store.md [hunk 5]
```diff
@@ -146,5 +146,304 @@ export default defineComponent({
...
+        const store = useStore()
+
+        return { store };
+    }
+});
+</script>
+```
+
+#### Using a typed store in Boot Files
+When using the store in Boot files it is also possible to use a typed store. Here is an example of a very simple boot file:
```
Comment (smolinari) [inline docs/src/pages/quasar-cli/vuex-store.md]: Comma correction

> When using the store in Boot files,

File: docs/src/pages/quasar-cli/vuex-store.md [hunk 6]
```diff
@@ -146,5 +146,304 @@ export default defineComponent({
...
+import { boot } from 'quasar/wrappers'
+import { root } from 'src/store'
+
+export default boot(({store}) => {
+    root.context(store).mutations.add(5)
+})
+```
+
+#### Using a typed store in Prefetch
+Similarly you can also use a typed store when using the [Prefetch feature](https://quasar.dev/quasar-cli/prefetch-feature). Here is an example:
```
Comment (smolinari) [inline docs/src/pages/quasar-cli/vuex-store.md]: Comma correction

> Similarly,

File: docs/src/pages/quasar-cli/vuex-store.md [hunk 7]
```diff
@@ -146,5 +146,304 @@ export default defineComponent({
...
+  state: ModuleState,
+  getters: ModuleGetters,
+  mutations: ModuleMutations,
+  actions: ModuleActions
+})
+
+export const useSplit = createComposable(split)
+```
+
+We then want to only load this module when a certain route component is visited. We can do that in (atleast) two different ways.
```
Comment (smolinari) [inline docs/src/pages/quasar-cli/vuex-store.md]: Comma correction

> We then want to only load this module,

> at least

File: docs/src/pages/quasar-cli/vuex-store.md [hunk 8]
```diff
@@ -146,5 +146,304 @@ export default defineComponent({
...
+
+The first method is using the [PreFetch Feature](/quasar-cli/prefetch-feature#store-code-splitting) that Quasar offers, similar as with regular Vuex. To do this we have a route defined in our `router/routes.ts` file, for this example we have a /split route which is a child of our MainLayout:
```
Comment (smolinari) [inline docs/src/pages/quasar-cli/vuex-store.md]: > similar to regular Vuex (?)

Comma correction

> To do this,

Better sentence structure? and Comma correction

> To do this, we have a route defined in our `router/routes.ts` file. For this example, we have a /split route which is a child of our MainLayout:

File: docs/src/pages/quasar-cli/vuex-store.md [hunk 9]
```diff
@@ -146,5 +146,304 @@ export default defineComponent({
...
+            unregisterModule($store, split)
+        })
+
+        return { greeting };
+    }
+});
+</script>
+```
+
+The second method is by using a `router.beforeEach` hook to register/ungregister our dynamic store modules. This makes sense if you have a separate store for lets say your admin panel of your website which is only used by a small percentage of visitors. You can then check if the route starts with `/admin` upon route navigation and load the store module based on that.
```
Comment (smolinari) [inline docs/src/pages/quasar-cli/vuex-store.md]: Comma correction

> This makes sense,

> separate store for, let's say, your admin panel of your website, which is only used ....

File: docs/src/pages/quasar-cli/vuex-store.md [hunk 10]
```diff
@@ -146,5 +146,304 @@ export default defineComponent({
...
+
+        return { greeting };
+    }
+});
+</script>
+```
+
+The second method is by using a `router.beforeEach` hook to register/ungregister our dynamic store modules. This makes sense if you have a separate store for lets say your admin panel of your website which is only used by a small percentage of visitors. You can then check if the route starts with `/admin` upon route navigation and load the store module based on that. 
+
+To do this you can use a [Boot File](/quasar-cli/boot-files) in Quasar that looks like this:
```
Comment (smolinari) [inline docs/src/pages/quasar-cli/vuex-store.md]: Comma correction

> To do this,

File: docs/src/pages/quasar-cli/vuex-store.md [hunk 11]
```diff
@@ -146,5 +146,304 @@ export default defineComponent({
...
+
+:::tip
+The example below is designed to work with both SSR and SPA. If you only use SPA this can be simplified by removing the last argument of `registerModule` entirely.
```
Comment (smolinari) [inline docs/src/pages/quasar-cli/vuex-store.md]: Comma correction

> If you only use SPA,

File: docs/src/pages/quasar-cli/vuex-store.md [hunk 12]
```diff
@@ -146,5 +146,304 @@ export default defineComponent({
...
+        } else {
+            if (store.hasModule('split'))
+                unregisterModule(store, split)
+        }
+        next()
+    })
+})
+```
+
+In your components you can then just use the dynamic module, without having to worry about registering it. For example:
```
Comment (smolinari) [inline docs/src/pages/quasar-cli/vuex-store.md]: Comma correction

> In your components,

**Comment by Smrtnyk** [comment]
Honestly to me this looks like a tutorial on how to use vuex smart modules
If they change their API this becomes obsolete.
Maybe there are already some tutorials we can link to rather than gardening something in here?

**Comment by Evertvdw** [comment]
@Smrtnyk True, but then specifically with Quasar and boot files etc.  The quasar implementation is always going to be a bit specific so I'm not aware of any existing documentation that we can refer to.

**Comment by Smrtnyk** [comment]
> @Smrtnyk True, but then specifically with Quasar and boot files etc. The quasar implementation is always going to be a bit specific so I'm not aware of any existing documentation that we can refer to.

I am not against these things, but what if for some reason you are not able to maintain this anymore and stuff change for that module or module goes to be deprecated
 I am afraid this could become obsolete and there is nobody to garden this documentation

**Comment by IlCallo** [comment]
@Smrtnyk since Pinia is the new official state management solution for Vue, there won't be any more Vuex major releases, thus vue-smart-module won't probably need any other API changes.

Gonna merge this for now, we can remove it when in a future major version when Vuex will be deprecated

**Pull request merged**